src/main/java/org/apache/maven/buildcache/checksum/AbstractInputAnalyzer.java

```diff
@@ -0,0 +1,262 @@
+    protected AbstractInputAnalyzer(
+            MavenProject project,
+            NormalizedModelProvider normalizedModelProvider,
+            ProjectInputCalculator projectInputCalculator,
+            MavenSession session,
+            CacheConfig config,
+            RepositorySystem repoSystem,
+            RemoteCacheRepository remoteCache,
+            ArtifactHandlerManager artifactHandlerManager,
+            String projectGlob,
+            ExclusionResolver exclusionResolver,
+            boolean processPlugins) {
+        this.project = project;
+        this.normalizedModelProvider = normalizedModelProvider;
+        this.projectInputCalculator = projectInputCalculator;
+        this.session = session;
+        this.config = config;
+        this.baseDirPath = project.getBasedir().toPath().toAbsolutePath();
+        this.repoSystem = repoSystem;
+        this.remoteCache = remoteCache;
+        this.projectGlob = projectGlob;
+        this.exclusionResolver = exclusionResolver;
+        this.processPlugins = processPlugins;
+        this.artifactHandlerManager = artifactHandlerManager;
+    }
+
+    /**
+     * Calculates the checksum for this analyzer's input type (source or test).
+     * This is the main entry point that subclasses must implement.
+     */
+    public abstract String calculateChecksum() throws IOException;
+
+    /**
+     * Gets the input files for this analyzer's type (source or test).
+     * Subclasses must implement to specify which directories and resources to scan.
+     */
+    protected abstract SortedSet<Path> getInputFiles() throws IOException;
+
+    /**
+     * Gets the dependencies for this analyzer's type (source or test).
+     * Subclasses must implement to specify which dependencies to include.
+     */
+    protected abstract SortedMap<String, String> getDependencies() throws IOException;
+
+    /**
+     * Gets the name of this analyzer type for logging purposes.
+     */
+    protected abstract String getAnalyzerType();
+
+    /**
+     * Common implementation for calculating checksum that uses the abstract methods.
+     */
+    protected String calculateChecksumInternal() throws IOException {
+        final long t0 = System.currentTimeMillis();
+
+        final SortedSet<Path> inputFiles = isPom(project) ? new TreeSet<>() : getInputFiles();
+        final SortedMap<String, String> dependenciesChecksum = getDependencies();
+
+        final long t1 = System.currentTimeMillis();
+
+        // Create a hash with enough capacity for all files and dependencies
+        // Each dependency contributes 2 items (key + value), each file contributes 1 item
+        final int totalItems = inputFiles.size() + (dependenciesChecksum.size() * 2);
+        final HashChecksum checksum = config.getHashFactory().createChecksum(totalItems);
+
+        // Add all input files to the hash
+        for (Path inputFile : inputFiles) {
+            checksum.update(inputFile.toString().getBytes());
```

Comment:
`getBytes()` is using the platform's default charset, for portability between win/linux it is better to set encoding.